grakn-graql/src/test/java/ai/grakn/graql/internal/analytics/CorenessTest.java

```diff
@@ -0,0 +1,349 @@
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assume.assumeFalse;
+
+public class CorenessTest {
+    private static final String thing = "thingy";
+    private static final String anotherThing = "anotherThing";
+    private static final String related = "related";
+    private static final String veryRelated = "veryRelated";
+
+    private ConceptId entityId1;
+    private ConceptId entityId2;
+    private ConceptId entityId3;
+    private ConceptId entityId4;
+
+    public GraknSession session;
+
+    @ClassRule
+    public final static SessionContext sessionContext = SessionContext.create();
+
+    @Before
+    public void setUp() {
+        session = sessionContext.newSession();
+    }
+
+    @Test(expected = GraqlQueryException.class)
+    public void testKSmallerThan2_ThrowsException() {
+        try (GraknTx graph = session.open(GraknTxType.READ)) {
+            graph.graql().compute().coreness().minK(1).execute();
+        }
+    }
+
+    @Test
+    public void testOnEmptyGraph_ReturnsEmptyMap() {
+        try (GraknTx graph = session.open(GraknTxType.READ)) {
+            Map<Integer, Set<String>> result = graph.graql().compute().coreness().execute();
+            assertTrue(result.isEmpty());
+        }
+    }
+
+    @Test
+    public void testOnGraphWithoutRelationships_ReturnsEmptyMap() {
+        try (GraknTx graph = session.open(GraknTxType.WRITE)) {
+            graph.putEntityType(thing).addEntity();
+            graph.putEntityType(anotherThing).addEntity();
+            Map<Integer, Set<String>> result = graph.graql().compute().coreness().execute();
+            assertTrue(result.isEmpty());
+        }
+    }
+
+    @Test
+    public void testOnGraphWithTwoEntitiesAndTwoRelationships() {
+        try (GraknTx graph = session.open(GraknTxType.WRITE)) {
+            EntityType entityType = graph.putEntityType(thing);
+            Entity entity1 = entityType.addEntity();
+            Entity entity2 = entityType.addEntity();
+
+            Role role1 = graph.putRole("role1");
+            Role role2 = graph.putRole("role2");
+            entityType.plays(role1).plays(role2);
+            graph.putRelationshipType(related)
+                    .relates(role1).relates(role2)
+                    .addRelationship()
+                    .addRolePlayer(role1, entity1)
+                    .addRolePlayer(role2, entity2);
+
+            Role role3 = graph.putRole("role3");
+            Role role4 = graph.putRole("role4");
+            entityType.plays(role3).plays(role4);
+            graph.putRelationshipType(veryRelated)
+                    .relates(role3).relates(role4)
+                    .addRelationship()
+                    .addRolePlayer(role3, entity1)
+                    .addRolePlayer(role4, entity2);
+
+            Map<Integer, Set<String>> result = graph.graql().compute().coreness().execute();
+            assertTrue(result.isEmpty());
+        }
+    }
```

Comment:
Just trying to have a general understanding. I understand the default k is 2. So when executing coreness for what values of k are calculations performed? Single specific value (here the default I guess)? Range of values starting from min k?

Reply:
A vertex u has coreness c if it belongs to a c-core but not to any (c+1)-core.
In coreness query, we start with k = minK, keep persisting and updating coreness of vertices.